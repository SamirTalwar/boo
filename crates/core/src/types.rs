//! A representation of a value's type.
//!
//! Used for type-checking and valid program synthesis.

use std::borrow::Borrow;
use std::fmt::Display;
use std::sync::Arc;

/// An opaque wrapper around a type.
pub trait TypeRef: From<Type<Self>> + Display + Sized {}

/// A simple type wrapper that allows for cycles.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Monotype(pub Arc<Type<Self>>);

impl AsRef<Type<Self>> for Monotype {
    fn as_ref(&self) -> &Type<Self> {
        self.0.as_ref()
    }
}

impl From<Type<Self>> for Monotype {
    fn from(value: Type<Self>) -> Self {
        Self(Arc::new(value))
    }
}

impl Display for Monotype {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.0.fmt(f)
    }
}

impl TypeRef for Monotype {}

/// A type bound by forall quantifiers.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct Polytype {
    pub quantifiers: Vec<TypeVariable>,
    pub mono: Monotype,
}

impl Polytype {
    pub fn unquantified(mono: Monotype) -> Self {
        Polytype {
            quantifiers: vec![],
            mono,
        }
    }
}

impl Display for Polytype {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        if self.quantifiers.is_empty() {
            write!(f, "{}", self.mono)
        } else {
            write!(f, "âˆ€")?;
            for quantifier in self.quantifiers.iter() {
                write!(f, " {quantifier}")?;
            }
            write!(f, ". ")?;
            write!(f, "{}", self.mono)
        }
    }
}

/// The set of types.
#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type<Outer: TypeRef> {
    Integer,
    Function { parameter: Outer, body: Outer },
    Variable(TypeVariable),
}

impl<Outer: TypeRef> Type<Outer> {
    pub fn transform<NewOuter: TypeRef>(self, f: impl Fn(Outer) -> NewOuter) -> Type<NewOuter> {
        match self {
            Type::Integer => Type::Integer,
            Type::Function { parameter, body } => Type::Function {
                parameter: f(parameter),
                body: f(body),
            },
            Type::Variable(variable) => Type::Variable(variable),
        }
    }
}

impl<Outer: TypeRef> Display for Type<Outer> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Type::Integer => write!(f, "Integer"),
            Type::Function { parameter, body } => write!(f, "({parameter} -> {body})"),
            Type::Variable(variable) => write!(f, "{variable}"),
        }
    }
}

#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Hash)]
pub struct TypeVariable(pub Arc<String>);

impl TypeVariable {
    pub fn new(value: String) -> Self {
        Self(Arc::new(value))
    }
    pub fn new_from_str(value: &str) -> Self {
        Self::new(value.to_owned())
    }
}

impl Borrow<str> for TypeVariable {
    fn borrow(&self) -> &str {
        self.0.as_ref()
    }
}

impl Display for TypeVariable {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        self.0.fmt(f)
    }
}
