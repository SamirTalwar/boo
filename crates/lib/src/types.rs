#[derive(Debug, Clone, PartialEq, Eq)]
pub enum Type {
    Integer,
    Function {
        parameter: Option<Box<Type>>,
        body: Option<Box<Type>>,
    },
}
