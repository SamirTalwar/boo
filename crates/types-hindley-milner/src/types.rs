use boo_core::types::{Monotype, Polytype, Type, TypeVariable};

use crate::fresh::FreshVariables;
use crate::subst::Subst;

pub trait Types {
    fn free(&self) -> im::HashSet<TypeVariable>;

    fn substitute(&self, substitutions: &Subst, fresh: &mut FreshVariables) -> Self;
}

impl Types for Type<Monotype> {
    fn free(&self) -> im::HashSet<TypeVariable> {
        match self {
            Type::Integer => im::HashSet::new(),
            Type::Function { parameter, body } => parameter.free().union(body.free()),
            Type::Variable(variable) => im::hashset![variable.clone()],
        }
    }

    fn substitute(&self, substitutions: &Subst, fresh: &mut FreshVariables) -> Self {
        match self {
            Type::Integer => Type::Integer,
            Type::Function { parameter, body } => Type::Function {
                parameter: parameter.substitute(substitutions, fresh),
                body: body.substitute(substitutions, fresh),
            },
            Type::Variable(variable) => match substitutions.get(variable) {
                None => Type::Variable(variable.clone()),
                Some(t) => (*t.0).clone(),
            },
        }
    }
}

impl Types for Monotype {
    fn free(&self) -> im::HashSet<TypeVariable> {
        self.0.free()
    }

    fn substitute(&self, substitutions: &Subst, fresh: &mut FreshVariables) -> Self {
        self.0.substitute(substitutions, fresh).into()
    }
}

impl Types for Polytype {
    fn free(&self) -> im::HashSet<TypeVariable> {
        let quantifiers = self.quantifiers.iter().cloned().collect();
        self.mono.free().relative_complement(quantifiers)
    }

    fn substitute(&self, substitutions: &Subst, fresh: &mut FreshVariables) -> Self {
        let replacements = self
            .quantifiers
            .iter()
            .map(|q| (q.clone(), fresh.next()))
            .collect::<Vec<_>>();
        let new_quantifiers = replacements
            .iter()
            .map(|(_, after)| after.clone())
            .collect();
        let replacements_subst = replacements
            .into_iter()
            .map(|(before, after)| (before, Type::Variable(after).into()))
            .collect();
        Self {
            quantifiers: new_quantifiers,
            mono: self
                .mono
                .substitute(&replacements_subst, fresh)
                .substitute(substitutions, fresh),
        }
    }
}
