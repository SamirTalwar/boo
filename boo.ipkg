package boo

sourcedir = "proofs"

modules =
    Builtins
  , EvaluationReduction
  , Expression
  , Identifier
  , Native
  , Primitive
  , Test
