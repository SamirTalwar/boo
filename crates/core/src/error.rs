//! The set of possible interpretation errors.

use crate::span::Span;
use crate::types;

/// An alias for [`Result`][std::result::Result] with the error type fixed to
/// [`Error`].
pub type Result<T> = std::result::Result<T, Error>;

/// The set of possible interpretation errors.
///
/// This can be used with [`thiserror`] and [`miette`].
#[derive(Debug, Clone, PartialEq, thiserror::Error, miette::Diagnostic)]
pub enum Error {
    #[error("Unexpected token: {token}")]
    #[diagnostic(code(boo::lexer::unexpected_token))]
    UnexpectedToken {
        #[label("unexpected token")]
        span: Span,
        token: String,
    },

    #[error("Parse error: expected one of {expected_tokens:?}")]
    #[diagnostic(code(boo::parser::error))]
    ParseError {
        #[label("{}", expected_one_of(expected_tokens))]
        span: Span,
        expected_tokens: Vec<&'static str>,
    },

    #[error("Match expression without a base case")]
    #[diagnostic(code(boo::verifier::match_without_base_case))]
    MatchWithoutBaseCase {
        #[label("match expression requires a base case")]
        span: Option<Span>,
    },

    #[error("Could not unify types")]
    #[diagnostic(code(boo::type_checker::type_error))]
    TypeUnificationError {
        #[label("{left_type}")]
        left_span: Option<Span>,
        left_type: types::Monotype,
        #[label("{right_type}")]
        right_span: Option<Span>,
        right_type: types::Monotype,
    },

    #[error("Could not apply the function")]
    #[diagnostic(code(boo::evaluator::invalid_function_application))]
    InvalidFunctionApplication {
        #[label("invalid function")]
        span: Option<Span>,
    },

    #[error("Invalid primitive")]
    #[diagnostic(code(boo::evaluator::type_error))]
    InvalidPrimitive {
        #[label("invalid primitive")]
        span: Option<Span>,
    },

    #[error("Unknown variable: {name:?}")]
    #[diagnostic(code(boo::evaluator::unknown_variable))]
    UnknownVariable {
        #[label("unknown variable")]
        span: Option<Span>,
        name: String,
    },
}

fn expected_one_of(strings: &[&str]) -> String {
    match strings {
        [] => "<nothing>".to_string(),
        [a] => format!("expected {}", a),
        [a, b] => format!("expected {} or {}", a, b),
        _ => format!(
            "expected one of {}, or {}",
            strings[0..strings.len() - 1].join(", "),
            strings[strings.len() - 1]
        ),
    }
}
