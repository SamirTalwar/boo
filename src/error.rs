#[derive(Debug, PartialEq)]
pub enum BooError {
    ParseError(peg::error::ParseError<peg::str::LineCol>),
}

impl std::fmt::Display for BooError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::ParseError(error) => write!(f, "Parse error: {}", error),
        }
    }
}
