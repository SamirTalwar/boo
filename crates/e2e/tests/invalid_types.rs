use boo::error::{Error, Result};
use boo::types::{Type, TypeVariable};
use boo::*;

#[test]
fn test_rejects_incorrect_types() -> Result<()> {
    expect_error(
        "rejects_incorrect_types",
        "1 + (fn x -> 3)",
        Error::TypeUnificationError {
            left_span: Some((0..14).into()),
            left_type: Type::Function {
                parameter: Type::Integer.into(),
                body: Type::Integer.into(),
            }
            .into(),
            right_span: Some((5..14).into()),
            right_type: Type::Function {
                parameter: Type::Variable(TypeVariable::new_from_str("_3")).into(),
                body: Type::Integer.into(),
            }
            .into(),
        },
    )
}

#[test]
fn test_match_expressions_must_be_of_the_same_type() -> Result<()> {
    expect_error(
        "match_expressions_must_be_of_the_same_type",
        "match 0 { 1 -> 2; _ -> fn x -> x }",
        Error::TypeUnificationError {
            left_span: Some((15..16).into()),
            left_type: Type::Integer.into(),
            right_span: Some((23..32).into()),
            right_type: Type::Function {
                parameter: Type::Variable(TypeVariable::new_from_str("_1")).into(),
                body: Type::Variable(TypeVariable::new_from_str("_1")).into(),
            }
            .into(),
        },
    )
}

#[test]
fn test_unknown_variable() -> Result<()> {
    expect_error(
        "unknown_variable",
        "123 + xyz",
        Error::UnknownVariable {
            span: Some((6..9).into()),
            name: "xyz".to_string(),
        },
    )
}

#[test]
fn test_does_not_close_over_variables_out_of_scope() -> Result<()> {
    expect_error(
        "does_not_close_over_variables_out_of_scope",
        "let fun = (let one = 1 in fn param -> one + param + external) in let external = 2 in fun 3",
        Error::UnknownVariable {
            span: Some((52..60).into()),
            name: "external".to_string(),
        },
    )
}

fn expect_error(name: &str, program: &str, expected_error: Error) -> Result<()> {
    let ast = parse(program)?.to_core()?;
    insta::with_settings!({ description => program }, {
        insta::assert_debug_snapshot!(name.to_string() + "__parse", ast);
    });

    let result = boo_types_hindley_milner::w(&ast);
    assert_eq!(result, Err(expected_error));

    Ok(())
}
