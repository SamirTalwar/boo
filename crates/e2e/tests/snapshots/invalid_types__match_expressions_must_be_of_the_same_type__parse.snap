---
source: crates/e2e/tests/invalid_types.rs
description: "match 0 { 1 -> 2; _ -> fn x -> x }"
expression: ast
---
Expr {
    span: Some(
        Span {
            start: 0,
            end: 34,
        },
    ),
    expression: Match(
        Match {
            value: Expr {
                span: Some(
                    Span {
                        start: 6,
                        end: 7,
                    },
                ),
                expression: Primitive(
                    Integer(
                        Small(
                            0,
                        ),
                    ),
                ),
            },
            patterns: [
                PatternMatch {
                    pattern: Primitive(
                        Integer(
                            Small(
                                1,
                            ),
                        ),
                    ),
                    result: Expr {
                        span: Some(
                            Span {
                                start: 15,
                                end: 16,
                            },
                        ),
                        expression: Primitive(
                            Integer(
                                Small(
                                    2,
                                ),
                            ),
                        ),
                    },
                },
                PatternMatch {
                    pattern: Anything,
                    result: Expr {
                        span: Some(
                            Span {
                                start: 23,
                                end: 32,
                            },
                        ),
                        expression: Function(
                            Function {
                                parameter: Name(
                                    "x",
                                ),
                                body: Expr {
                                    span: Some(
                                        Span {
                                            start: 31,
                                            end: 32,
                                        },
                                    ),
                                    expression: Identifier(
                                        Name(
                                            "x",
                                        ),
                                    ),
                                },
                            },
                        ),
                    },
                },
            ],
        },
    ),
}
